independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/AbstractSharedContext.java

```diff
@@ -116,6 +116,10 @@ public boolean equals(Object obj) {
                 return false;
             }
             Key other = (Key) obj;
+            // Shortcut removes hotspot on contextual.equals
+            if (contextual == other.contextual) {
```

Comment:
Hm, at the moment we don't implement `equals()/hashCode()` for generated bean classes so `!contextual.equals(other.contextual)` should be translated to `!(contextual == other.contextual)`. In other words, this modification could save one `java.lang.Object.equals(Object)` invocation (which is very likely negligible although it removes the equals method from the hot path). `Long.toHexString()` saves probably a lot more because `String.format()` creates a formatter object, parses the string, etc.

Reply:
@mkouba what's you're saying is we should drop the equals() call altogether?

Reply:
Nope. I'd like to keep it. The fact that we don't implement it now does not mean we'll never need to implement it.

Reply:
OK. But if it ended up being a hot spot for Paul, there's something weird going on.

Reply:
It's hard to say without the app and test sources. CC @bcluap

Reply:
In JVM mode, JIT inlining will make the code path the same in no more than 1000 invocations (short method threshold) so this can not be an hotspot.
In native mode, I don't know if there is any inlining.
Anyway, method call is very chip so it may be a profiling artefact.
Some profilers inject bytecode so inlining didn't happens and we see these kind of wrong hotspot.
This kind of hack can go wrong at some moment so I suggest dropping it.
Side question, which profiler was used ? Asynch-profiler didn't inject bytecode so it may not show this line as an hotspot.

Reply:
Yep, I've removed the `Key` completely in the follow-up PR https://github.com/quarkusio/quarkus/pull/11217.

Reply:
So, I just saw that the profiler used is the one from VisualVM, it inject some bytecode and can generates deoptimization.
That's why it can misleading you to such kind of hotspot.
When working on lowlevel optimizations (and avoiding a method call is very low level), you should use low level stuff like async-profiler, we have a page that expalain how to use it: https://github.com/quarkusio/quarkus/blob/master/TROUBLESHOOTING.md

So again, the hotspot for the equals method call is certainly a profiler artefact not a real hotspot. 
And 1% difference between two load testing run is very slight so can go into the recording error range.
Such small performance enhancements should be validated via microbenchmarking using JMH ...

Reply:
> Such small performance enhancements should be validated via microbenchmarking using JMH ...

Yes, we talked about this in the comments below, e.g. https://github.com/quarkusio/quarkus/pull/11148#issuecomment-668564910.

Reply:
If you have many possible concrete types implementing `Contextual`, the dispatch to find the right `equals` implementation becomes a very expensive megamorphic call.
+1 for the shortcut

Reply:
As mentioned above we removed the need for `equals()` from the hot path completely...

Reply:
> As mentioned above we removed the need for `equals()` from the hot path completely...

Sure that's even better. I just meant to suggest a possible explanation for the equals to be - in some contexts - really not that efficient. Removing the field is even better!